Rework Book (#1230)
graphql-rust/juniper: book/src/types/objects/complex_fields.md

Complex fields
==============

> Using a plain [Rust struct][struct] for representing a [GraphQL object][0] is easy and trivial, but what if need to express something non-trivial behind a [GraphQL field][4]? For example:

Reviewer: ```suggestion
Using a plain [Rust struct][struct] for representing a [GraphQL object][0] is easy and trivial but does not cover every case. What if you need to express something non-trivial as a [GraphQL field][4], such as:
```

> - Perform some non-trivial logic while [executing][1] the [field][4] (like querying database, etc.).

Reviewer: ```suggestion
- Calling non-trivial logic while [executing][1] the [field][4] (like querying database, etc.).
```

> - Accept [field arguments][5].

Reviewer: ```suggestion
- Accepting [field arguments][5].
```

> - Define a circular [GraphQL object][0], where one of its [fields][4] returns the type itself.

Reviewer: ```suggestion
- Defining a circular [GraphQL object][0], where one of its [fields][4] returns the type itself.
```

> - Use other [Rust] type (non-[struct]) to represent a [GraphQL object][0].

Reviewer: ```suggestion
- Using some other (non-[struct]) [Rust] type to represent a [GraphQL object][0].
```

> To express this, we need a way to define a [GraphQL field][4] as a function. In [Juniper] this is achievable by placing the [`#[graphql_object]` attribute][3] on the [`impl` block][6], which turn its methods into [GraphQL fields][4]:

Reviewer: ```suggestion
To support these more complicated usecases, we need a way to define a [GraphQL field][4] as a function. In [Juniper] this is achievable by placing the [`#[graphql_object]` attribute][3] on an [`impl` block][6], which turn its methods into [GraphQL fields][4]:
```

    inhabitants: Vec<Person>,
}

// Defines the `House` GraphQL object.
#[graphql_object]
impl House {
    // Creates the field `inhabitantWithName(name: String!)`, 
    // returning a `null`able `Person`.
    fn inhabitant_with_name(&self, name: String) -> Option<&Person> {
        self.inhabitants.iter().find(|p| p.name == name)
    }
}
#
# fn main() {}
```
> **NOTE**: To access global data such as database connections or authentication information, a _context_ is used. To learn more about this, see the ["Context" chapter](context.md).

### Default arguments

> Though [Rust] doesn't have notion of default arguments, [GraphQL arguments][4] are able to have default values (the one used when a GraphQL operation doesn't specify them explicitly). In [Juniper], defining a default value for a [GraphQL argument][4] is possible with the `#[graphql(default)]` attribute:

Reviewer: ```suggestion
Though [Rust] doesn't have the notion of default arguments, [GraphQL arguments][4] are able to have default values. These default values are used when a GraphQL operation doesn't specify the argument explicitly. In [Juniper], defining a default value for a [GraphQL argument][4] is enabled by the `#[graphql(default)]` attribute:
```

We can override the name by using the `#[graphql(name = "...")]` attribute:
```rust
# extern crate juniper;
# use juniper::graphql_object;
#
struct Person;

#[graphql_object]
#[graphql(name = "PersonObject")]
impl Person { // exposed as `PersonObject` in GraphQL schema
    #[graphql(name = "myCustomFieldName")]
    fn renamed_field( // exposed as `myCustomFieldName` in GraphQL schema
        #[graphql(name = "myArgument")]
        renamed_argument: bool, // exposed as `myArgument` in GraphQL schema
    ) -> bool {
        renamed_argument
    }
}
#
# fn main() {}
```

Or provide a different renaming policy for all the defined [fields][4]:
```rust
# extern crate juniper;
# use juniper::graphql_object;
#
struct Person;

#[graphql_object]
#[graphql(rename_all = "none")] // disables any renaming
impl Person {
    fn renamed_field( // exposed as `renamed_field` in GraphQL schema
        renamed_argument: bool, // exposed as `renamed_argument` in GraphQL schema
    ) -> bool {
        renamed_argument
    }
}
#
# fn main() {}
```
> **TIP**: Supported policies are: `SCREAMING_SNAKE_CASE`, `camelCase` and `none` (disables any renaming).

### Documentation and deprecation

> Similarly, [GraphQL fields][4] could also be [documented][7] and [deprecated][9] via `#[graphql(description = "...")]` and `#[graphql(deprecated = "...")]`/[`#[deprecated]`][13] attributes:

Reviewer: ```suggestion
Similarly, [GraphQL fields][4] may also be [documented][7] and [deprecated][9] via `#[graphql(description = "...")]` and `#[graphql(deprecated = "...")]`/[`#[deprecated]`][13] attributes:
```

#[graphql_object]
#[graphql(description = "This description overwrites the one from doc comment.")]
impl Person {
    /// This doc comment is visible only in Rust API docs.
    #[graphql(description = "This description is visible only in GraphQL schema.")]
    fn empty() -> &'static str {
        ""
    }
    
    #[graphql(desc = "This description is visible only in GraphQL schema.")]
    //        ^^^^ shortcut for a `description` argument
    fn field(
        #[graphql(desc = "This description is visible only in GraphQL schema.")]
        arg: bool,
    ) -> bool {
        arg
    }

    /// This doc comment is visible in both Rust API docs and GraphQL schema 
    /// descriptions.
    #[graphql(deprecated = "Just because.")]
    fn deprecated_graphql() -> bool {
        true
    }
    
    // Standard Rust's `#[deprecated]` attribute works too!
    #[deprecated(note = "Reason is optional, btw!")]
    fn deprecated_standard() -> bool { // has no description in GraphQL schema
        false
    }
}
#
# fn main() {}
```
> **NOTE**: Only [GraphQL object][0]/[interface][11] fields and [GraphQL enum][10] values can be [deprecated][9].

### Ignoring

> By default, all methods of an [`impl` block][6] are considered as [GraphQL fields][4]. If a helper method is needed, then it should be either defined in a separate [`impl` block][6], or marked with the `#[graphql(ignore)]` attribute:

Reviewer: ```suggestion
By default, all methods of an [`impl` block][6] are exposed as [GraphQL fields][4]. If a method should not be exposed as a [GraphQL field][4], it should be defined in a separate [`impl` block][6] or marked with the `#[graphql(ignore)]` attribute:
```

```rust
# #![allow(dead_code)]
# extern crate juniper;
# use juniper::graphql_object;
#
struct Person {
    name: String,
    age: i32,
}

#[graphql_object]
impl Person {
    fn name(&self) -> &str {
        self.name.as_str()
    }

    fn age(&self) -> i32 {
        self.age
    }

    #[graphql(ignore)]
    pub fn hidden_from_graphql(&self) {
        // whatever goes...
    }

    #[graphql(skip)]
    //        ^^^^ alternative naming, up to your preference
    pub fn also_hidden_from_graphql(&self) {
        // whatever goes...
    }
}

impl Person {
    pub fn not_even_considered_for_graphql(&self) {
        // whatever goes...
    }
}
#
# fn main() {}
```

> > **TIP**: See more features in API docs of the [`#[graphql_object]`][3] attribute.

Reviewer: ```suggestion
> **TIP**: See more features in the API docs of the [`#[graphql_object]`][3] attribute.
```
